include/kodlab_mjbots_sdk/pid_controller.h

```diff
@@ -0,0 +1,533 @@
+   * @return Returns the Pid output
+   */
+  VectorNS Update(Scalar state,Scalar derivative_state) {
+    //checking if using for scalar PID
+    static_assert(N==1, "Using Scalar function in vector PID");
+    // converts scalar to vector form to calculate output
+    return Update(state*VectorNS::Ones(),derivative_state*VectorNS::Ones());
+  }
+  /**
+   * @brief Function to update the PID output given the respective errors
+   * @param state Current state for the controller.
+   * @param derivative_state Current derivative state of the controller.  
+   * @return Returns the Pid output
+   */
+  VectorNS Update(const VectorNS & state,const VectorNS & derivative_state) {
+
+    return Update(state,derivative_state,time_step_);
+  }
+
+  /**
+   * @brief Function to update the PID output given the respective errors
+   * @param error_p Proportional Error
+   * @param error_d Derivative Error
+   * @return Returns the Pid output
+   */
+  
+  virtual VectorNS UpdateWithError(const VectorNS & error_p, const VectorNS &  error_d, double time_step) {
+    // Computing the integral error
+    i_error_ += error_p*time_step;
+
+    //adding a limit on integral error to prevent windup
+    i_error_ = i_error_.cwiseMin(windup_limit_).cwiseMax(-windup_limit_);
+
+    // Finding the control output from the errors
+    output_ = kp_.array() * error_p.array() + kd_.array() * error_d.array() + ki_.array() * i_error_.array();
+
+    //Checking if in Deadband range
+    for(int i=0;i<error_p.rows();i++){
+      if ( std::abs(error_p.coeff(i)) < deadband_.coeff(i)) {
+          output_(i) =0; 
+          i_error_(i)=0;
+      } 
+    }
+    
+    return output_;
+  }
+
+  /**
+   * @brief Destructor
+   */
+  ~PIDController() {}
+
+
+protected:
+  
+  
+  /**
+   * @brief State Setpoint 
+   */
+  //Eigen::Matrix<Scalar,Eigen::Dynamic,1> setpoint_;
+  VectorNS setpoint_;
+
+  /**
+   * @brief State derivative Setpoint
+   */
+  VectorNS derivative_setpoint_;
+  /**
+   * @brief Time difference between 2 updates
+   */
+  double time_step_;
+
+  /**
+   * @brief Proportional Gain
+   */
+  VectorNS kp_;
+
+  /**
+   * @brief Derivative Gain
+   */
+  VectorNS kd_;
+
+  /**
+   * @brief Integral Gain
+   */
+  VectorNS ki_;
+
+  /**
+   * @brief Stores the previous error for the d term
+   */
+  VectorNS d_error_ ;
+
+  /**
+   * @brief Sums up the error for calculating the I term
+   */
+  VectorNS i_error_ ;
+
+  /**
+   * @brief Error at each step.
+   */
+  VectorNS error_ ;
+
+  /**
+   * @brief Deadband of the system
+   */
+  VectorNS deadband_;
+
+  /**
+   * @brief Windup limit of the system if needed
+   */
+  VectorNS windup_limit_;
+  
+  /**
+   * @brief Max control output
+   */
+  std::array<Scalar, 2> saturation_;
+
+  /**
+   * @brief Output of the function used internally
+   */
+  VectorNS output_ ;
+
+  static constexpr int kSaturationLowerLimitIndex = 0;
+  static constexpr int kSaturationUpperLimitIndex = 1;
```

Comment:
maybe add  the following replacing all the 0.001 default arguments
`static constexpr float kDefaultTimeStep = 0.001;`